File: python/leetcodehard/minimum_window_substring.py
"""
https://leetcode.com/problems/minimum-window-substring/

status - completed
"""

from collections import Counter

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def minWindow(self, s, t):
        if not s or not t:
            return ""

        need = Counter(t)
        missing = len(t)
        left = 0
        best_left, best_right = 0, 0

        for right, char in enumerate(s, 1):
            if need[char] > 0:
                missing -= 1
            need[char] -= 1

            if missing == 0:
                while need[s[left]] < 0:
                    need[s[left]] += 1
                    left += 1

                if best_right == 0 or right - left < best_right - best_left:
                    best_left, best_right = left, right

        return s[best_left:best_right]


if __name__ == "__main__":
    test_input_1 = ("ADOBECODEBANC", "ABC")
    output = Solution().minWindow(*test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ("a", "a")
    output = Solution().minWindow(*test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/minimumWindowSubstring.ts
/*
https://leetcode.com/problems/minimum-window-substring/

status - completed
*/

import { logp } from "../utils/logger";

function minWindow(s: string, t: string): string {
  if (t.length === 0 || s.length < t.length) return "";

  const need = new Map<string, number>();
  for (const ch of t) {
    need.set(ch, (need.get(ch) ?? 0) + 1);
  }

  let required = need.size;
  let formed = 0;
  const windowCounts = new Map<string, number>();

  let left = 0;
  let bestLen = Infinity;
  let bestStart = 0;

  for (let right = 0; right < s.length; right++) {
    const ch = s[right];
    windowCounts.set(ch, (windowCounts.get(ch) ?? 0) + 1);

    if (need.has(ch) && windowCounts.get(ch) === need.get(ch)) {
      formed++;
    }

    while (formed === required) {
      if (right - left + 1 < bestLen) {
        bestLen = right - left + 1;
        bestStart = left;
      }

      const leftCh = s[left];
      windowCounts.set(leftCh, windowCounts.get(leftCh)! - 1);
      if (need.has(leftCh) && windowCounts.get(leftCh)! < need.get(leftCh)!) {
        formed--;
      }
      left++;
    }
  }

  return bestLen === Infinity ? "" : s.slice(bestStart, bestStart + bestLen);
}

logp(`input s="ADOBECODEBANC" t="ABC" expected BANC actual ${minWindow("ADOBECODEBANC", "ABC")}`);
logp(`input s="a" t="a" expected a actual ${minWindow("a", "a")}`);
logp(`input s="a" t="aa" expected "" actual "${minWindow("a", "aa")}"`);
